vividus-plugin-rest-api/src/main/java/org/vividus/http/HttpRequestExecutor.java

```diff
@@ -53,52 +55,53 @@ public HttpRequestExecutor(IHttpClient httpClient, HttpTestContext httpTestConte
      * Executes HTTP request and frees context request entity and headers.
      * Request is executed safely for unchecked exceptions: HttpRequestBuildException, ConnectionClosedException.
      * Assertion fails if exceptions occurred.
-     * @param httpMethod HttpMethod to execute
-     * @param endpoint Request endpoint
-     * @param relativeURL Relative URL - optional
-     * @throws IOException If an input or output exception occurred
+     * @param requestData HTTP request input data
      */
-    public void executeHttpRequest(HttpMethod httpMethod, String endpoint, Optional<String> relativeURL)
-            throws IOException
+    public void executeHttpRequest(HttpRequestData requestData)
     {
-        executeHttpRequest(httpMethod, endpoint, relativeURL, response -> false);
+        executeHttpRequest(requestData, response -> false, Duration.ZERO, 1);
     }
 
     /**
      * Executes HTTP request <i>while predicate condition is true</i>
      * and frees context request entity and headers afterwards.
      * Request is executed safely for unchecked exceptions: HttpRequestBuildException, ConnectionClosedException.
      * Assertion fails if exceptions occurred.
-     * @param httpMethod HttpMethod to execute
-     * @param endpoint Request endpoint
-     * @param relativeURL Relative URL - optional
+     * @param requestData HTTP request input data
      * @param repeatRequest Predicate condition to check response
-     * @throws IOException If an input or output exception occurred
+     * @param duration Time duration to wait
+     * @param retryTimes How many times request will be retried; duration/retryTimes=timeout between requests
      */
-    public void executeHttpRequest(HttpMethod httpMethod, String endpoint, Optional<String> relativeURL,
-            Predicate<HttpResponse> repeatRequest) throws IOException
+    public void executeHttpRequest(HttpRequestData requestData, Predicate<HttpResponse> repeatRequest,
+            Duration duration, int retryTimes)
     {
         try
         {
-            HttpResponse httpResponse;
-            do
-            {
-                httpResponse = executeHttpCallSafely(httpMethod, endpoint, relativeURL.orElse(null));
-            }
-            while (repeatRequest.test(httpResponse));
+            Supplier<HttpResponse> responseProvider = () -> {
+                try
+                {
+                    return executeHttpCallSafely(requestData);
+                }
+                catch (IOException e)
+                {
+                    throw new HttpRequestException("An I/O exception occurred", e);
```

Comment:
what's the reason to wrap exception? it makes their handling non-transparent

Reply:
Return to a problem with method which throws checked exception and is used in Supplier construction (line 90). Don't know how to avoid now. Custom Supplier looks tricky too.

Reply:
@valfirst  Can we start to store them into some utility package? :)